### PR TITLE
Fix potential race condition in cloud-provider

### LINKED ISSUE #70: Prevent Empty ProviderID in CloudNodeLifecycleController
**What happened**:

When creating new nodes the existence of the property `node.Spec.ProviderID` is not sure. Due to race conditions, the `ProviderID` may not yet be set by the CloudProviderImplemtation, but the node is already evaluated by CCM.                                                                                                                                                        
Currently this results in an empty ProviderID provided to the CloudProviderImplemtation by the CloudLifecycleControlle when calling `shutdownInCloudProvider`. If the implementation is not prepared to work with an empty ProviderID, it will cause an error in the implementation.

**What you expected to happen**:

The `node.Spec.ProviderID` should be set when the controller is able to fetch the correct ProviderID from the provider implementation.

The CloudLifecycleController itself is prepared to work with an empty ProviderID, but does not store the evaluated ProviderID in the node reference used later on (see [ensureNodeExistsByProviderID](https://github.com/kubernetes/cloud-provider/blob/a5b00cde736d87324a85853929b83eddd855f49f/controllers/nodelifecycle/node_lifecycle_controller.go#L231C2-L245C3)). 

A simple assignment of the evaluated ProviderID will fix the issue. 

**How to reproduce it**:

This is a race condition and cannot be reproduced easily.
It occurs randomly and can be kind of reproduced with clusters that have more than 10 worker nodes.

**Anything else we need to know?**:

We fixed this already in our fork succesfully.

<sup>Michael Grote [michael.grote@mercedes-benz.com](mailto:tobias.giese@mercedes-benz.com), Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/mercedes-benz/foss/blob/master/LEGAL_IMPRINT.md)</sup>

### PR DESCRIPTION
**Pull request by tobiasgiese**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR sets the `providerID` to the node pointer to have it later.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/cloud-provider/issues/70

#### Special notes for your reviewer:

There is a potential race condition in the cloud-provider during a node join:

* T0 controller A (node controller) writes Node data (i.e., `Node.Spec.ProviderID`)
* T1 conrtoller B (lifecycle controller) reads the old Node data
* T2 controller B cannot read the ProviderID and returns with an error

We are validating if the node exists with `ensureNodeExistsByProviderID` and if `node.Spec.ProviderID` is not set we are fetching the `instanceID` from the cloud provider.
https://github.com/kubernetes/kubernetes/blob/f3996039ef13ef75b0c91a7329f5098851e5ec34/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L230-L245
Later we are calling `shutdownInCloudProvider` which needs the `ProviderID` as well, but we are not fetching it from the cloud provider ID if it's unset
https://github.com/kubernetes/kubernetes/blob/f3996039ef13ef75b0c91a7329f5098851e5ec34/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L210

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

<sub>Tobias Giese <tobias.giese@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/mercedes-benz/foss/blob/master/LEGAL_IMPRINT.md)</sub>

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go [hunk 1]
```diff
@@ -242,6 +242,7 @@ func ensureNodeExistsByProviderID(ctx context.Context, cloud cloudprovider.Inter
 			klog.Warningf("Cannot find valid providerID for node name %q, assuming non existence", node.Name)
 			return false, nil
 		}
+		node.Spec.ProviderID = providerID
```
Comment (tobiasgiese) [inline staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:245]: Alternatively we can fetch the ProviderID from the cloud provider again, like already implemented in the func [`ensureNodeExistsByProviderID`](https://github.com/kubernetes/kubernetes/blob/f3996039ef13ef75b0c91a7329f5098851e5ec34/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L231-L245). But to reduce API calls against the provider we should simply adjust the pointer as proposed in this PR.

**Comment by dims** [comment]
cc @mckdev @mmerkes

**Comment by dims** [comment]
cc @mmerkes @nckturner

**Comment by mmerkes** [comment]
Tx for putting up the fix. Although it's maybe a little unexpected to set the field as part of that call, it seems harmless and a simple solution. Is this masking a potential bug in the node registration flow where the `providerID` should be set and it's not?

**Comment by tobiasgiese** [comment]
> Is this masking a potential bug in the node registration flow where the `providerID` should be set and it's not?

I just tested this again on our OpenStack environment with a 15 worker nodes test cluster. And I can say that it looks like tit produces a bug.
Normally the providerID should look like `openstack:///instance-id`, but in some cases (where the bug we are trying to fix occurs) the providerID was set to `/instance-id`.

So fetching the providerID again (as proposed in https://github.com/kubernetes/kubernetes/pull/121383#discussion_r1366688659) would be the best way. WDYT @mmerkes?

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go [hunk 2]
```diff
@@ -207,7 +207,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
 		return false, errors.New("cloud provider does not support instances")
 	}
 
-	shutdown, err := instances.InstanceShutdownByProviderID(ctx, node.Spec.ProviderID)
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
```
Comment (tobiasgiese) [inline staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:211]: We could add this block into a helper function.

**Comment by aojea** [comment]
can we add a test that reproduces the bug and covers future regressions?

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go [hunk 3]
```diff
@@ -207,7 +207,24 @@ func shutdownInCloudProvider(ctx context.Context, cloud cloudprovider.Interface,
...
+		var err error
+		providerID, err = instances.InstanceID(ctx, types.NodeName(node.Name))
```
Comment (mmerkes) [inline staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: Isn't [getProviderID](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L434) the method you should be using here?

Comment (tobiasgiese) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: Maybe it's a duplicated implementation. It is just copied from `ensureNodeExistsByProviderID`  which is in the same file

https://github.com/kubernetes/kubernetes/blob/5ab5300a3aae230162b676eda4520f81c0923115/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L237C6-L263

Comment (tobiasgiese) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: But the main difference is that we are ignoring the error `InstanceNotFound` which we should as we are in the function `shutdownInCloudProvider`. This should not happen if it is already deleted.

**Comment by tobiasgiese** [comment]
> can we add a test that reproduces the bug and covers future regressions?

Don't know if this is somehow easily possible without changing the fake cloud. It's just a implementation detail inside `shutdownInCloudProvider()`. Or do you have an idea how to achieve this @aojea?

**Comment by tobiasgiese** [comment]
> I just tested this again on our OpenStack environment with a 15 worker nodes test cluster. And I can say that it looks like it produces a bug.
> Normally the providerID should look like `openstack:///instance-id` (Edit: while looking into the Kubernetes Node spec.), but in some cases (where the bug we are trying to fix occurs) the providerID was set to `/instance-id` w/o the prefix.

This could be fixed by setting the provider name as prefix, like in `GetProviderID()`

```go
node.Spec.ProviderID = cloud.ProviderName() + "://" + providerID
```

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go (hunk 3 above)
Comment (mmerkes) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: Ah, I see. However, looks like the result is going to be different between call `instances.InstanceID(ctx, types.NodeName(node.Name))` and `cnc.getProviderID(ctx, copyNode)`, right?

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/cloud.go#L101-L115

Comment (mmerkes) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: Ah, looks like you discuss in another comment. Seems like a bug that it's using `instances.InstanceID(ctx, types.NodeName(node.Name))` anywhere to get `providerID`?

Comment (tobiasgiese) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: I think we should use both. Get the provider ID via `getProviderID` and ignore the error if no instance is found.
Will update the PR tomorrow

Comment (mmerkes) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: Makes sense. If we're getting the provider ID, we should get it properly.

**Comment by aojea** [comment]
didn't check this PR in deep, but you have an example here https://github.com/kubernetes/kubernetes/pull/117249#issuecomment-1505636060

**Comment by aojea** [comment]
Can you explain where is the race? when those two function run in parallel?

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go (hunk 3 above)
Comment (tobiasgiese) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go:213]: One problem: the function `GetInstanceProviderID` is returning a different error to the const `InstanceNotFound` if the instance is not found

`InstanceNotFound`
https://github.com/kubernetes/kubernetes/blob/2f270bd99609561be53893d3c7fc77d8efe7ac4e/staging/src/k8s.io/cloud-provider/cloud.go#L256

general error of `GetInstanceProviderID`
https://github.com/kubernetes/kubernetes/blob/2f270bd99609561be53893d3c7fc77d8efe7ac4e/staging/src/k8s.io/cloud-provider/cloud.go#L112

But I will change the code that a real provider ID is used

**Comment by tobiasgiese** [comment]
> Can you explain where is the race? when those two function run in parallel?

Tbh, I don't understand I am not 100% sure at the moment, but I think that the race is between the node controller (in `syncNode()`) and the node lifecycle controller (in `MonitorNodes()`).
If a new node joins the cluster `syncNode()` will add the `Node.Spec.ProviderID` to the Node, but `MonitorNodes()` is referring to this value at the same time.

File: staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go (hunk 3 above)
Comment (tobiasgiese) [reply staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go]: I created a method `getProviderID`, WDYT?

**Comment by aojea** [comment]
> Tbh, I am not 100% sure at the moment. I think that the race is between the node controller (in `syncNode()`) and the node lifecycle controller (in `MonitorNodes()`).

two controllers can not race, both need to operate in their own view of the current state and not share dependencie ... the reconcile loops will make them eventually consistent, this is by design

**Comment by tobiasgiese** [comment]
> two controllers can not race, both need to operate in their own view of the current state and not share dependencie ... the reconcile loops will make them eventually consistent, this is by design

That's true, but it's a kind of a race if they are consuming the same resource. Do we need a test in this case? IMO it's totally fair to fix it as it is.

**Comment by aojea** [comment]
> Tbh, I am not 100% sure at the moment. I think that the race is between the node controller (in syncNode()) and the node lifecycle controller (in MonitorNodes()).

> That's true, but it's a kind of a race if they are consuming the same resource. 

can you explain the race in terms of 

- T0 controllerX do 
- T1 controller Y do

we need to be 100% sure of the problem we are solving before merging, that is why I asked for the test reproducing the issue, I'm not disagreeing there can be a race, but we need to be able to reproduce it

**Comment by tobiasgiese** [comment]
> can you explain the race in terms of
> 
> * T0 controllerX do
> * T1 controller Y do

I tried to explain it in this comment https://github.com/kubernetes/kubernetes/pull/121383#issuecomment-1774050807.

> If a new node joins the cluster, `syncNode()` will add the `Node.Spec.ProviderID` to the Node, but `MonitorNodes()` is referring to this value at the same time.

`syncNode()` is adding the value `Node.Spec.ProviderID` but `MonitorNodes()` is also trying to ready it.

**Comment by aojea** [comment]
chatted offline and clarified the misunderstanding, I also remembered there was another PR addressing this issue and I found out it was not merged, this is a duplicate of https://github.com/kubernetes/kubernetes/pull/120615/,  isn't it?

**Comment by tobiasgiese** [comment]
> this is a duplicate of #120615, isn't it?

it is a dup, yes. Closing the issue. Thanks for your help @aojea

**Pull request closed without merging**